server/src/main/java/org/opensearch/index/shard/ReplicationGroup.java

```diff
@@ -75,7 +81,9 @@ public ReplicationGroup(
                 skippedShards.add(shard);
             } else {
                 if (trackedAllocationIds.contains(shard.allocationId().getId())) {
-                    replicationTargets.add(shard);
+                    replicationTargets.add(
+                        new ReplicationModeAwareShardRouting(replicationModeMap.get(shard.allocationId().getId()), shard)
```

Comment:
Shouldn't we check that `replicationModeMap.get(shard.allocationId().getId())` has any non-null value?

Reply:
`CheckpointState` object always has a non-null `replicationMode`. Here, `replicationModeMap` is constructed out of `checkpoints` within `ReplicationTracker` which has all `trackedAllocationIds` (which is a subset of keys of `checkpoints`).

However, I have added a non-null check in `CheckpointState` constructor for `replicationMode`.